Add support for Kafka's ProducerConfig.CONNECTIONS_MAX_IDLE_MS_CONFIG (#1090)

MarvinCai opened the pull request:
Motivation:
Support configure Kafka's ProducerConfig.CONNECTIONS_MAX_IDLE_MS_CONFIG in PulsarKafkaProducer by utilizing pulsar's ClientBuilder.keepAliveInterval.

MarvinCai commented:
rerun java8 tests
rerun integration tests
rerun cpp tests

sijie commented on pulsar-client-kafka-compat/pulsar-client-kafka/src/main/java/org/apache/kafka/clients/producer/PulsarKafkaProducer.java:
```diff
@@ -114,7 +114,11 @@ private PulsarKafkaProducer(Map<String, Object> conf, Properties properties, Ser
+            // Support Kafka's ProducerConfig.CONNECTIONS_MAX_IDLE_MS_CONFIG in ms.
+            long keepAliveIntervalMs = Long.parseLong(properties.getProperty(ProducerConfig.CONNECTIONS_MAX_IDLE_MS_CONFIG, "30000"));
+            // If passed in value is greater than Integer.MAX_VALUE in second will throw ArithmeticException.
+            int keepAliveInterval = Math.toIntExact(keepAliveIntervalMs / 1000);
```
can you catch `ArithmeticException` and log an error message (e.g. "Invalid CONNECTIONS_MAX_IDLE_MS_CONFIG value - " + keepAliveIntervalMs + ". Please use a value smaller than " + Integer.MAX_VALUE") and then throw an InvalidArgumentException with the error message?

MarvinCai replied:
Done, catch ArithmeticException, log error message and throw IllegalArgumentException with same error message.

sijie commented on pulsar-client-kafka-compat/pulsar-client-kafka/src/test/java/org/apache/kafka/clients/producer/PulsarKafkaProducerTest.java:
```diff
@@ -0,0 +1,48 @@
+        properties.put(ProducerConfig.BOOTSTRAP_SERVERS_CONFIG, Arrays.asList("pulsar://localhost:6650"));
+        properties.put(ProducerConfig.CONNECTIONS_MAX_IDLE_MS_CONFIG, "1000000");
+
+        PulsarKafkaProducer<String, String> pulsarKafkaProducer = new PulsarKafkaProducer<>(properties, null, null);
```
Can you verify mockClientBuilder.keepAliveInterval is called once with the provided value?

MarvinCai replied:
Verified mockClientBuilder.keepAliveInterval is called exactly once with expected value.

MarvinCai commented:
rerun java8 tests

MarvinCai commented:
Addressed comments.

sijie commented on pulsar-client-kafka-compat/pulsar-client-kafka/src/main/java/org/apache/kafka/clients/producer/PulsarKafkaProducer.java:
```diff
@@ -34,6 +34,7 @@
 import java.util.concurrent.TimeoutException;
 import java.util.stream.Collectors;
 
+import com.sun.javaws.exceptions.InvalidArgumentException;
```
it seems that this is a wrong import?

MarvinCai replied:
Removed unused import.
Also added document that merlimat mentioned.

MarvinCai commented:
rerun java8 tests
rerun integration tests
rerun cpp tests

MarvinCai commented:
rerun java8 tests
rerun integration tests
rerun cpp tests

MarvinCai commented:
rerun java8 tests

MarvinCai commented:
rerun java8 tests

MarvinCai commented:
> This should be also updated in the consumer configuration and also we need to update the table with the supported configs in `site2/docs/adaptors-kafka.md`
Updated the description for this config in the table.

MarvinCai commented:
rerun java8 tests

MarvinCai commented:
rerun java8 tests

MarvinCai commented:
rerun java8 tests

MarvinCai commented:
rerun java8 tests

sijie commented:
@merlimat can you review this PR again?

MarvinCai commented:
rerun java8 tests
rerun integration tests

MarvinCai commented:
rerun integration tests

MarvinCai commented:
rerun integration tests

sijie commented:
@MarvinCai thank you for your contribution!

MarvinCai commented:
> @MarvinCai thank you for your contribution!
Thank you @sijie and @merlimat for reviewing the code and help me with this pr.
Will carry on and try contribute more  :smiley:
